### PR TITLE
Specifying name for transpiler output(Issue #2745)

### PR DESCRIPTION
**Pull request by IsidoreCarinae**
### Summary
If specified in the function call, the output circuits can now have unique names.
Attempting to resolve:
https://github.com/Qiskit/qiskit-terra/issues/2745

### CODE REVIEW

**Comment by 1ucian0** [comment]
Hi @IsidoreCarinae ! Thanks a lot for your contribution. It seems that there is a problem with the default parameter (when no name is provided) Also, run `make lint` `make style` to make sure that the code follows our style guidelines. You can run the tests locally by `make tests`.

Thanks!

**Comment by IsidoreCarinae** [comment]
@1ucian0 Hi, I made a commit today for fixes in naming functionality and added a test. Please lemme know if more changes need to be made. The code is not linted*

File: test/python/transpiler/test_naming_transpiled_circuits.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Comment (1ucian0) [inline test/python/transpiler/test_naming_transpiled_circuits.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: test/python/transpiler/test_naming_transpiled_circuits.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        qr3 = QuantumRegister(2)
+        cr3 = ClassicalRegister(2)
+        circ3 = QuantumCircuit(qr3, cr3)
+        circ3.x(qr3[0])
+        circ3.x(qr3[1])
+        # combining multiple circuits
+        circuits = [circ1, circ2, circ3]
+        backendie = BasicAer.get_backend('qasm_simulator')
+        # equal lengths
+        names = ['awesome_circ1', 'awesome_circ2', 'awesome_circ3']
```
Comment (maddy-tod) [inline test/python/transpiler/test_naming_transpiled_circuits.py]: These names use underscores but when you check there are hypens, is this translation done somewhere?

Comment (kdk) [reply test/python/transpiler/test_naming_transpiled_circuits.py]: Good catch @maddy-tod , the assertions below are `self.assertTrue`'s but should probably be `self.assertEqual`s.

Comment (1ucian0) [reply test/python/transpiler/test_naming_transpiled_circuits.py]: Awesome catch! The full method is not running because the `test_` prefix is missed!

Comment (1ucian0) [reply test/python/transpiler/test_naming_transpiled_circuits.py]: I fixed it in dac91196

Comment (1ucian0) [reply test/python/transpiler/test_naming_transpiled_circuits.py]: I fixed all the tests. Please, have a look!

**Comment by 1ucian0** [comment]
Hi @maddy-tod, @IsidoreCarinae, and @kdk 

I made several changes in the tests. Can you please have a look?

File: qiskit/compiler/transpile.py [hunk 3]
```diff
@@ -191,12 +191,41 @@ def callback_func(**kwargs):
...
+                                      'in {} '.format(circuit.name) +
                                       'is greater than maximum ({}) '.format(max_qubits) +
                                       'in the coupling_map')
     # Transpile circuits in parallel
     circuits = parallel_map(_transpile_circuit, list(zip(circuits, transpile_configs)))
-
+    # naming and returning circuits
+    # output_names could be either a string or a list
+    if output_names is not None:
+        if isinstance(output_names, str):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: fixed in 21d87d0a

File: qiskit/compiler/transpile.py [hunk 4]
```diff
@@ -191,12 +191,41 @@ def callback_func(**kwargs):
...
+            # single circuit
+            if len(circuits) == 1:
+                circuits[0].name = output_names
+            # multiple circuits
+            else:
+                raise TranspilerError("Expected a list object of length equal " +
+                                      "to that of the number of circuits " +
+                                      "being transpiled")
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Docstring updated in 5f91c828

File: qiskit/compiler/transpile.py [hunk 5]
```diff
@@ -31,7 +31,7 @@ def transpile(circuits,
               basis_gates=None, coupling_map=None, backend_properties=None,
               initial_layout=None, seed_transpiler=None,
               optimization_level=None,
-              pass_manager=None, callback=None):
+              pass_manager=None, callback=None, output_names=None):
```
[the remark this answers is not in the document]
Comment (1ucian0) [reply qiskit/compiler/transpile.py]: Agree. Renamed in 63cac558

**Pull request merged**